File: src/siminov/connect/reader/QuickServiceDescriptorReader.java
/** 
 * [SIMINOV FRAMEWORK]
 * Copyright [2015] [Siminov Software Solution LLP|support@siminov.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

package siminov.connect.reader;

import java.io.IOException;
import java.io.InputStream;
import java.util.Iterator;

import org.xml.sax.Attributes;
import org.xml.sax.SAXException;

import siminov.connect.Constants;
import siminov.connect.model.ApplicationDescriptor;
import siminov.connect.model.ServiceDescriptor;
import siminov.connect.resource.ResourceManager;
import siminov.core.exception.PrematureEndOfParseException;
import siminov.core.exception.SiminovException;
import siminov.core.log.Log;
import siminov.core.reader.SiminovSAXDefaultHandler;
import android.content.Context;

/**
 * Exposes methods to quickly parse service descriptor defined by application.
 */
public class QuickServiceDescriptorReader extends SiminovSAXDefaultHandler implements Constants {

	private StringBuilder tempValue = new StringBuilder();
	private String finalServiceDescriptorName = null;
	
	private Context context = null;
	
	private ServiceDescriptor serviceDescriptor = null;
	
	private boolean doesMatch = false;
	private boolean isNameProperty = false;
	
	private siminov.core.resource.ResourceManager coreResourceManager = siminov.core.resource.ResourceManager.getInstance();
	private ResourceManager connectResourceManager = ResourceManager.getInstance();
	
	public QuickServiceDescriptorReader(final String findServiceDescriptorName) throws SiminovException {
		
		if(findServiceDescriptorName == null || findServiceDescriptorName.length() <= 0) {
			Log.error(getClass().getName(), "Constructor", "Invalid Service Descriptor Name Which Needs To Be Searched.");
			throw new SiminovException(getClass().getName(), "Constructor", "Invalid Service Descriptor Name Which Needs To Be Searched.");
		}
		
		this.finalServiceDescriptorName = findServiceDescriptorName;
	}
	
	public void process() throws SiminovException {
		context = coreResourceManager.getApplicationContext();
		if(context == null) {
			Log.error(getClass().getName(), "process", "Invalid Application Context found.");
			throw new SiminovException(getClass().getName(), "process", "Invalid Application Context found.");
		}


		
		ApplicationDescriptor applicationDescriptor = connectResourceManager.getApplicationDescriptor();
		Iterator<String> serviceDescriptorPaths = applicationDescriptor.getServiceDescriptorPaths();
		
		while(serviceDescriptorPaths.hasNext()) {
			String serviceDescriptorPath = serviceDescriptorPaths.next();
			
			InputStream databaseMappingDescriptorStream = null;
			try {
				databaseMappingDescriptorStream = context.getAssets().open(serviceDescriptorPath);
			} catch(IOException ioException) {
				Log.error(getClass().getName(), "process", "IOException caught while getting input stream of Service Descriptor: " + serviceDescriptorPath + ", " + ioException.getMessage());
				throw new SiminovException(getClass().getName(), "process", "IOException caught while getting input stream of Service Descriptor: " + serviceDescriptorPath + ", " + ioException.getMessage());
			}
			
			try {
				parseMessage(databaseMappingDescriptorStream);
			} catch(Exception exception) {
				Log.error(getClass().getName(), "process", "Exception caught while parsing Service Descriptor: " + serviceDescriptorPath + ", " + exception.getMessage());
				throw new SiminovException(getClass().getName(), "process", "Exception caught while parsing Service Descriptor: " + serviceDescriptorPath + ", " + exception.getMessage());
			}
			
			if(doesMatch) {

				ServiceDescriptorReader serviceDescriptor = new ServiceDescriptorReader(serviceDescriptorPath);
				this.serviceDescriptor = serviceDescriptor.getServiceDescriptor();
				
				return;
			}
		}
	}
	
	public void startElement(String uri, String localName, String qName, Attributes attributes) throws SAXException {
		
		tempValue = new StringBuilder();

		if(localName.equalsIgnoreCase(Constants.SERVICE_DESCRIPTOR_PROPERTY)) {
			String propertyName = attributes.getValue(SERVICE_DESCRIPTOR_PROPERTY_NAME);

			if(propertyName.equalsIgnoreCase(SERVICE_DESCRIPTOR_NAME)) {
				isNameProperty = true;
			}
		} 
	}
	
	public void characters(char[] ch, int start, int length) throws SAXException {
		String value = new String(ch,start,length);
		
		if(value == null || value.length() <= 0 || value.equalsIgnoreCase(siminov.core.Constants.NEW_LINE)) {
			return;
		}
		
		value = value.trim();
		tempValue.append(value);
	}

	public void endElement(String uri, String localName, String qName) throws SAXException {
		
		if(localName.equalsIgnoreCase(Constants.SERVICE_DESCRIPTOR_PROPERTY)) {
			
			if(isNameProperty) {
				if(tempValue.toString().equalsIgnoreCase(finalServiceDescriptorName)) {
					doesMatch = true;
				}
				
				throw new PrematureEndOfParseException(getClass().getName(), "endElement", "Service Descriptor Name: " + tempValue);
			}
		}
	}


	/**
	 * Get service descriptor
	 * @return Service Descriptor
	 */
	public ServiceDescriptor getServiceDescriptor() {
		return this.serviceDescriptor;
	}
	
	/**
	 * Set service descriptor
	 * @return Service Descriptor
	 */
	public boolean containServiceDescriptor() {
		return this.doesMatch;
	}
}

File: Windows/universal8.1/Siminov/Connect/Reader/QuickServiceDescriptorReader.cs
/** 
 * [SIMINOV FRAMEWORK]
 * Copyright [2015] [Siminov Software Solution LLP|support@siminov.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/



using Siminov.Connect.Model;
using Siminov.Connect.Resource;
using Siminov.Core.Exception;
using Siminov.Core.Log;
using Siminov.Core.Reader;
using Siminov.Core.Utils;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml;

namespace Siminov.Connect.Reader
{
    public class QuickServiceDescriptorReader : SiminovSAXDefaultHandler
    {

	    private StringBuilder tempValue = new StringBuilder();
	    private String finalServiceDescriptorName = null;
		
	    private ServiceDescriptor serviceDescriptor = null;
	
	    private bool doesMatch = false;
	    private bool isNameProperty = false;
	
	    private Core.Resource.ResourceManager coreResourceManager = Core.Resource.ResourceManager.GetInstance();
	    private ResourceManager connectResourceManager = ResourceManager.GetInstance();
	
	    public QuickServiceDescriptorReader(String findServiceDescriptorName) 
        {
		
		    if(findServiceDescriptorName == null || findServiceDescriptorName.Length <= 0) 
            {
			    Log.Error(this.GetType().Name, "Constructor", "Invalid Service Descriptor Name Which Needs To Be Searched.");
			    throw new SiminovException(this.GetType().Name, "Constructor", "Invalid Service Descriptor Name Which Needs To Be Searched.");
		    }
		
		    this.finalServiceDescriptorName = findServiceDescriptorName;
	    }
	
	    public void Process() 
        {
		
		    ApplicationDescriptor applicationDescriptor = connectResourceManager.GetApplicationDescriptor();
		    IEnumerator<String> serviceDescriptorPaths = applicationDescriptor.GetServiceDescriptorPaths();
		
		    while(serviceDescriptorPaths.MoveNext()) 
            {
			    String serviceDescriptorPath = serviceDescriptorPaths.Current;

                String serviceDescriptorFileName = null;
                String serviceDescriptorFilePath = null;

                serviceDescriptorFilePath = serviceDescriptorPath.Substring(0, serviceDescriptorPath.LastIndexOf("/"));
                serviceDescriptorFileName = serviceDescriptorPath.Substring(serviceDescriptorPath.LastIndexOf("/") + 1, (serviceDescriptorPath.Length - serviceDescriptorPath.LastIndexOf("/")) - 1);

                Stream serviceDescriptorStream = null;

                try 
                {
                    serviceDescriptorStream = FileUtils.SearchFile(serviceDescriptorFilePath, serviceDescriptorFileName, FileUtils.INSTALLED_FOLDER);
                } 
                catch(System.Exception ioException) 
                {
                    Log.Error(this.GetType().Name, "Process", "IOException caught while getting input stream of Service Descriptor: " + serviceDescriptorPath + ", " + ioException.Message);
                    throw new SiminovException(this.GetType().Name, "process", "IOException caught while getting input stream of Service Descriptor: " + serviceDescriptorPath + ", " + ioException.Message);
                }
			
			    try 
                {
                    ParseMessage(serviceDescriptorStream);
			    }
                catch (PrematureEndOfParseException exception) 
                {
                    Log.Error(this.GetType().Name, "Process", "PrematureEndOfParseException caught while parsing Service Descriptor: " + serviceDescriptorPath + ", " + exception.Message);
			    }
			
			    if(doesMatch) 
                {

				    ServiceDescriptorReader serviceDescriptor = new ServiceDescriptorReader(serviceDescriptorPath);
				    this.serviceDescriptor = serviceDescriptor.GetServiceDescriptor();
				
				    return;
			    }
		    }
	    }
	
        public override void StartElement(XmlReader reader, IDictionary<String, String> attributes)
        {		
		
            String localName = reader.Name;
		    tempValue = new StringBuilder();

		    if(localName.Equals(Constants.SERVICE_DESCRIPTOR_PROPERTY)) 
            {
			    String propertyName = attributes[Constants.SERVICE_DESCRIPTOR_PROPERTY_NAME];

			    if(propertyName.Equals(Constants.SERVICE_DESCRIPTOR_NAME, StringComparison.OrdinalIgnoreCase)) 
                {
				    isNameProperty = true;
			    }
		    } 
	    }
	
        public override void Characters(String value)
        {
            if (value == null || value.Length <= 0 || value.Equals(Core.Constants.NEW_LINE, StringComparison.OrdinalIgnoreCase))
            {
                return;
            }

            value = value.Trim();
            tempValue.Append(value);
	    }

        public override void EndElement(String localName)
        {
		
		    if(localName.Equals(Constants.SERVICE_DESCRIPTOR_PROPERTY, StringComparison.OrdinalIgnoreCase)) 
            {
			
			    if(isNameProperty) 
                {
				    if(tempValue.ToString().Equals(finalServiceDescriptorName, StringComparison.OrdinalIgnoreCase)) 
                    {
					    doesMatch = true;
				    }
				
				    throw new PrematureEndOfParseException(this.GetType().Name, "EndElement", "Service Descriptor Name: " + tempValue);
			    }
		    }
	    }



        /// <summary>
        /// Get service descriptor
        /// </summary>
        /// <returns>Service Descriptor</returns>
	    public ServiceDescriptor GetServiceDescriptor() 
        {
		    return this.serviceDescriptor;
	    }
	

        /// <summary>
        /// Set service descriptor
        /// </summary>
        /// <returns>Service Descriptor</returns>
	    public bool ContainServiceDescriptor() 
        {
		    return this.doesMatch;
	    }

    }
}
